### DISCUSSION TITLE
How to set Contract Address in Ton Jetton transaction ?

### DISCUSSION THREAD

**Discussion by devanshu6198** [trustwallet/wallet-core#4179] [Q&A] [answered]
### Discussed in https://github.com/trustwallet/wallet-core/discussions/4171

<div type='discussions-op-text'>

<sup>Originally posted by **devanshu6198** December 16, 2024</sup>
![Screenshot 2024-12-16 at 16 35 16](https://github.com/user-attachments/assets/212fd9c0-3121-43fd-ba23-3276a2fdb5a6)

Where can i put contract address ?</div>

**Comment by satoshiotomakan** [comment, marked as answer]
Hi @devanshu6198, you don't need to specify the jetton master address explicitly, but your jetton wallet address as `SigningInput.dest`. Please note that `JettonTransfer.toOwner` should be a recipient's jetton wallet address as well.

To convert your main TON address to a jetton wallet address, you can use `TONConverter` module as at the example:
```kt
     val mainAddress = "UQBjKqthWBE6GEcqb_epTRFrQ1niS6Z1Z1MHMwR-mnAYRoYr"  // Your main TON address
     val mainAddressBoc = TONAddressConverter.toBoc(mainAddress)
     assertEquals(mainAddressBoc, "te6cckEBAQEAJAAAQ4AMZVVsKwInQwjlTf71KaItaGs8SXTOrOpg5mCP004DCNAptHQU") 
  
     // Call `get_wallet_address` of the NOT Jetton contract (`EQAvlWFDxGF2lXm67y4yzC17wYKD9A0guwPkMs1gOsM__NOT`):
     // curl --location 'https://toncenter.com/api/v2/runGetMethod' --header 'Content-Type: application/json' --data \ 
     // '{"address":"EQAvlWFDxGF2lXm67y4yzC17wYKD9A0guwPkMs1gOsM__NOT","method":"get_wallet_address","method":"get_wallet_address","stack":[["tvm.Slice","te6cckEBAQEAJAAAQ4AMZVVsKwInQwjlTf71KaItaGs8SXTOrOpg5mCP004DCNAptHQU"]]}' 
  
     // Parse the `get_wallet_address` RPC response. 
     val jettonAddressBocEncoded = "te6cckEBAQEAJAAAQ4AFvT5rqwxcbKfITqnkwL+go4Zi9bulRHAtLt4cjjFdK7B8L+Cq" 
     val jettonAddress = TONAddressConverter.fromBoc(jettonAddressBocEncoded) 
     assertEquals(jettonAddress, "UQAt6fNdWGLjZT5CdU8mBf0FHDMXrd0qI4FpdvDkcYrpXV5H") 
```

**Comment by devanshu6198** [comment]
Hi @satoshiotomakan 
 
- ![Screenshot 2024-12-17 at 18 34 09](https://github.com/user-attachments/assets/5b4ea337-bf09-46e5-a099-651ef29f5769)

 Here My Transaction is Continue failing 
 
 toOwner  = "UQAyi5L7thM59vwciZTIuhnD2WPOjmdmBSJcWb0xDpujdnpa" (Receiver).
 TonAddress Convert  = "te6cckEBAQEAJAAAQ4AOBgowR7qL+5SzXSIwe1UvEHCQDvZucQcmjIHKfGMXizAxSGVT" with Receiver Address.
 Get New Boc Address = "te6cckEBAQEAJAAAQ4AamGnhrCnhkmvHItjXmC+/GyVwF9rj7cqtsVe9ZuwNyRAy1k2c"
 responseAddress = "UQDUw08NYU8Mk145Fsa8wX342SuAvtcfblVtir3rN2BuSJVg"
 setDest = "EQCxE6mUtQJKFnGfaROTKOt1lZbDiiX1kCixRv7Nw2Id_sDs"
 
 it is correct ?

**Comment by satoshiotomakan** [comment]
`dest` should be your jetton wallet address, but EQCxE6mUtQJKFnGfaROTKOt1lZbDiiX1kCixRv7Nw2Id_sDs is a jetton master (contract) address.
Let's say the code snippet above is a function `fn getJettonWalletAddr(main_addr: String) -> String`. Then
```kt
val myJettonAddr = getJettonWalletAddr(myTonAddr)
val toOwnerJettonAddr = getJettonWalletAddr(toOwnerTonAddr)

val jettonTransfer = TheOpenNetwork.JettonTransfer.newBuilder()
            .setJettonAmount(500 * 1000 * 1000)
            .setToOwner(toOwnerJettonAddr)
            .setResponseAddress(myTonAddr)
            .setForwardAmount(1)
            .build()

val transfer = TheOpenNetwork.Transfer.newBuilder()
            .setDest(myJettonAddr)
            .setAmount(100 * 1000 * 1000)
            .setMode(TheOpenNetwork.SendMode.PAY_FEES_SEPARATELY_VALUE or TheOpenNetwork.SendMode.IGNORE_ACTION_PHASE_ERRORS_VALUE)
            .setComment("test comment")
            .setBounceable(true)
            .setJettonTransfer(jettonTransfer)
```

**Reply by devanshu6198** [reply]
@satoshiotomakan 

I am Following Every Step For Both Side Sender And Receiver Address for Jetton Wallet But Every Transaction Are Fail.

**Reply by satoshiotomakan** [reply]
Can you try to set the recipient's TON address instead of Jetton Wallet address?

**Reply by satoshiotomakan** [reply]
Here's how we create the JettonTransfer signing message:

```kt
        val asset = transaction.asset
        val fromAddress = asset.account.address

        val bocCell = fromAddress.toBoc()
        val bocTvm = rpcClient.getJettonsAddress(transaction.asset.contract!!, bocCell)
        val jettonAddress = bocTvm.fromBoc().toUserFriendly(bounceable = true)

        val forwardAmount = when (walletInfo.await().result.isActive) {
            true -> 1L // if > 0, will send notification message to origin wallet
            false -> 0L // Do not send notification, destination wallet can't do it as it is inactive
        }

        val jettonsTransfer = baseTransferSigningInput.copy(
            amount = maxOf(fee.amount.longValue(), RecommendedTonAmountForJettonTransfer.longValue()),
            mode = SendMode.PAY_FEES_SEPARATELY.value or SendMode.IGNORE_ACTION_PHASE_ERRORS.value,
            dest = jettonAddress, // sender's jetton wallet
            jetton_transfer = JettonTransfer(
                jetton_amount = transaction.amount.longValue(),
                response_address = fromAddress, // send change output coins back to sender
                to_owner = transaction.to.toUserFriendly(bounceable = true),
                forward_amount = forwardAmount,
            ),
        )
```

**Comment by devanshu6198** [comment]
@satoshiotomakan 

Is there any method for first time send jetton token to new receiver address ?
or 
Is there any method for sending first time jetton token to receiver address ?

**Comment by satoshiotomakan** [comment]
@devanshu6198 I think there is no need to handle new receiver address specifically. Please test it carefully before delivering to the production

**Comment by devanshu6198** [comment]
@satoshiotomakan 

- https://tonviewer.com/transaction/b2f1dba5d37452c7a928042394accf2826d516d9a49ed9fc4de6f822c4845aee

Can please you tell me what is problem in this transaction ?

**Comment by satoshiotomakan** [comment]
JettonTransfer's destination is 0:328b92fbb61339f6fc1c8994c8ba19c3d963ce8e676605225c59bd310e9ba376, however it's a TON wallet, not a Jetton wallet

**Reply by devanshu6198** [reply]
@satoshiotomakan 

Not A jetton Wallet.

- https://tonviewer.com/transaction/a322fc343163433c2deb10b8d3e1b210248e83257dfc7e1dc3879a8a51dbf079 

 Again create it with Jetton wallet but it is also fail.

**Reply by satoshiotomakan** [reply]
It remains the same. You need to use `val toOwnerJettonAddr = getJettonWalletAddr(toOwnerTonAddr)` address at `JettonTransfer.setToOwner(toOwnerJettonAddr)`

**Reply by devanshu6198** [reply]
Yes, You Are Right .

I have add same as you say in previous discussion, But not working.

**Reply by devanshu6198** [reply]
@satoshiotomakan 

![Screenshot 2024-12-18 at 17 00 39](https://github.com/user-attachments/assets/ea1b7dd6-c5d1-4ac3-bbca-b68d1734caed)

toOwner in data get from to the BOC.
responseAddress is a sender address.
setDest in the sender jetton wallet address.

it is correct ?

**Comment by devanshu6198** [comment]
@satoshiotomakan 

Thank You for your support
